user_sessions/backends/db.py

```diff
@@ -19,89 +13,46 @@ def __init__(self, session_key=None, user_agent=None, ip=None):
         self.ip = ip
         self.user_id = None
 
+    @classmethod
+    def get_model_class(cls):
+        # Avoids a circular import and allows importing SessionStore when
+        # user_sessions is not in INSTALLED_APPS
+        from ..models import Session
+
+        return Session
+
     def __setitem__(self, key, value):
         if key == auth.SESSION_KEY:
             self.user_id = value
         super().__setitem__(key, value)
 
-    def load(self):
-        try:
-            s = Session.objects.get(
-                session_key=self.session_key,
-                expire_date__gt=timezone.now()
-            )
-            self.user_id = s.user_id
-            # do not overwrite user_agent/ip, as those might have been updated
-            if self.user_agent != s.user_agent or self.ip != s.ip:
-                self.modified = True
-            return self.decode(s.session_data)
-        except (Session.DoesNotExist, SuspiciousOperation) as e:
-            if isinstance(e, SuspiciousOperation):
-                logger = logging.getLogger('django.security.%s' %
-                                           e.__class__.__name__)
-                logger.warning(force_str(e))
-            self.create()
-            return {}
-
-    def exists(self, session_key):
-        return Session.objects.filter(session_key=session_key).exists()
+    def _get_session_from_db(self):
+        s = super()._get_session_from_db()
+        self.user_id = s.user_id
```

Comment:
`self` here is the `SessionStore` instance. Are you sure that settings the `user_id` for it is correct?

Reply:
This function is used by Django's [`load`](https://github.com/django/django/blob/7bb741d787ba360a9f0d490db92e22e0d28204ed/django/contrib/sessions/backends/db.py#L41) method, and it's [what our previous `load` method did](https://github.com/jazzband/django-user-sessions/blob/5ab2c57df315b3351557e559951ed0df9df78a8b/user_sessions/backends/db.py#L33).

Reply:
Why does the `SessionStore` keep a `user_id` like this? Are there guarantees that a single `SessionStore` is not shared between concurrent threads?

Please don't consider this question a blocking problem, I'm mostly trying to ensure that I understand this correctly.

Reply:
> Please don't consider this question a blocking problem, I'm mostly trying to ensure that I understand this correctly.

All good, me too. We're in this together! 😄

My best guess is that Django's own SessionStore is not supposed to be shared between threads at all, since the `django.contrib.sessions.backends.base.SessionBase` [stores a few attributes, including `_session_key`, on itself](https://github.com/django/django/blob/7bb741d787ba360a9f0d490db92e22e0d28204ed/django/contrib/sessions/backends/base.py#L44):

```python
    def __init__(self, session_key=None):
        self._session_key = session_key
        self.accessed = False
        self.modified = False
        self.serializer = import_string(settings.SESSION_SERIALIZER)
```

Since the upstream code seems to be happy to share an entire session key on an instance, I assume it's safe to also share a `user_id` attribute.

Reply:
And in our previous code, `self.user_id` is only used in the [`save` method](https://github.com/jazzband/django-user-sessions/blob/5ab2c57df315b3351557e559951ed0df9df78a8b/user_sessions/backends/db.py#L75):

```python
    def save(self, must_create=False):
        """
        Saves the current session data to the database. If 'must_create' is
        True, a database error will be raised if the saving operation doesn't
        create a *new* entry (as opposed to possibly updating an existing
        entry).
        """
        obj = Session(
            session_key=self._get_or_create_session_key(),
            session_data=self.encode(self._get_session(no_load=must_create)),
            expire_date=self.get_expiry_date(),
            user_agent=self.user_agent,
            user_id=self.user_id,  # <------- HERE
            ip=self.ip,
        )
        using = router.db_for_write(Session, instance=obj)
        try:
            with transaction.atomic(using):
                obj.save(force_insert=must_create, using=using)
        except IntegrityError as e:
            if must_create and 'session_key' in str(e):
                raise CreateError
            raise
```

In this PR code, I use it in `create_model_instance`:

```python
    # Used in DBStore.save()
    def create_model_instance(self, data):
        """
        Return a new instance of the session model object, which represents the
        current session state. Intended to be used for saving the session data
        to the database.
        """
        return self.model(
            session_key=self._get_or_create_session_key(),
            session_data=self.encode(data),
            expire_date=self.get_expiry_date(),
            user_agent=self.user_agent,
            user_id=self.user_id,
            ip=self.ip,
        )
```

which is called by the [superclass' save method](https://github.com/django/django/blob/7bb741d787ba360a9f0d490db92e22e0d28204ed/django/contrib/sessions/backends/db.py#L73):

```python
    def save(self, must_create=False):
        """
        Save the current session data to the database. If 'must_create' is
        True, raise a database error if the saving operation doesn't create a
        new entry (as opposed to possibly updating an existing entry).
        """
        if self.session_key is None:
            return self.create()
        data = self._get_session(no_load=must_create)
        obj = self.create_model_instance(data)  # <--- HERE
        using = router.db_for_write(self.model, instance=obj)
        try:
            with transaction.atomic(using=using):
                obj.save(
                    force_insert=must_create, force_update=not must_create, using=using
                )
        except IntegrityError:
            if must_create:
                raise CreateError
            raise
        except DatabaseError:
            if not must_create:
                raise UpdateError
            raise
```

Reply:
Thanks